upcoming: [M3-7354] DC Get Well - Disabled regions

abailly-akamai opened the pull request:
## Description 📝
This PR implements the UI for disabled regions in the `<RegionSelect />`. We originally had hardcoded "fake" regions  to represent disabled menu items. DC Get Well now implements a new endpoint (`account/availability`) which represents an array of DCs (or regions) unavailable to the user. 

The UI is essentially the same (disabled item plus tooltip with link to a static info page - link TBD).

Right now this PR does not implement the behavior in all `RegionSelect` instances to keep the PR small and focus on pure logic implementation. (see testing steps)

## Changes  🔄
- Implement new `account/availability` request (mocked for now) and underlying logic to determine the region's availability to a given user.
- Implement a new prop to pass the current capability to the region select
- remove all logic pertaining to fake regions 🎉 
- cleanup and enhanced test suites

## Preview 📷

![Screenshot 2023-11-21 at 13 35 40](https://github.com/linode/manager/assets/130582365/34f08b4d-6b2f-4248-a578-f484fa2bdd6e)

## How to test 🧪

### Prerequisites
- Pull code locally
- Turn on the DC Get Well feature flag in dev settings
- Turn on MSW

### Verification steps 
- You can verify the behavior by adding the current capability prop (you can try for various create flows)

ex: add `currentCapability="Block Storage"` to the [VolumeCreate region select](https://github.com/linode/manager/blob/743c345141708cfcb851f5f16d00f1ed8c0609b0/packages/manager/src/features/Volumes/VolumeCreate.tsx#L293) 

👉  verify that Atlanta and Tokyo are disabled items

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -146,6 +167,8 @@ export const RegionSelect = React.memo((props: RegionSelectProps) => {
         groupBy={(option: RegionSelectOption) => option.data.region}
         label={label ?? 'Region'}
+        loading={accountAvailabilityLoading}
+        loadingText="Loading regions..."
```
Adding loading props since we now have an API request that may affect the rendering. It's unlikely to be noticable at all and i doubt we'll ever see the loading state but does not hurt to plan for it.

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.types.ts:
```diff
@@ -16,6 +21,9 @@ export interface RegionSelectProps
   > {
+  // TODO DC_GET_WELL
+  // Make this prop required & remove the undefined union type once all consumers are updated
+  currentCapability?: Capabilities | undefined;
```
Please note this the comment above. The prop IS going to be required when we wrap up this implementation. But for the sake of keeping the PR small and not add the prop on every select (there's some slight extra work to do for some of our services) I kept it optional. right now.

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.types.ts:
```diff
@@ -25,3 +33,23 @@ export interface RegionSelectProps
+  accountAvailabilityData: AccountAvailability[] | undefined;
+  // TODO DC_GET_WELL
+  // remove the undefined union type once all consumers are updated
+  currentCapability: Capabilities | undefined;
```
same here

abailly-akamai commented on packages/manager/src/components/RegionSelect/disabledRegions.tsx:
buh-bye! 🎉

coliu-akamai reviewed (approved):
✅  Confirmed current capability prop works as expected on flows 
✅ general code review

Good test additions + I liked how you passed in objects to the region utility functions, definitely helped with readability!

abailly-akamai commented:
VPC test failure looks unrelated. It passed locally every time I ran it locally

abailly-akamai commented:
@coliu-akamai can you please give the PR another run after this fix? wanna make sure we don't have any regressions - thx!

coliu-akamai commented:
thank you @hana-linode for catching this!!! 🙏 

taking another look through it now

hana-linode commented on packages/manager/src/components/RegionSelect/RegionSelect.styles.ts:
```diff
@@ -1,16 +1,25 @@
+    textTransform: 'initial',
+  },
+  '& .MuiAutocomplete-listbox': {
+    '& li:first-child .MuiAutocomplete-groupLabel': {
```
```suggestion
    '& li:first-of-type .MuiAutocomplete-groupLabel': {
```

![image](https://github.com/linode/manager/assets/115299789/5f9c194a-a423-453c-b6c3-619cf2d5752c)

abailly-akamai replied:
right on - fixed!

coliu-akamai commented:
- Confirmed keyboard accessibility works now! (had one question - for disabled items' tooltip, is there anyway to make that available for keyboard and voiceover users?)
- tests pass
- region select still works as expected with capabilities

@hana-linode looked into this a bit - Miami wasn't in the cached regions.json data that the MSW uses, so it also didn't show up for me here. I added 'Block Storage' to some other regions that are in regions.json and they're disabled 😄 
![image](https://github.com/linode/manager/assets/139280159/0ac3eaf8-a83c-4aad-8bd1-b933dc4ae9a2)

does anyone know how the regions.json file gets generated? If Miami isn't consistently included in that file, could we switch Miami out for a region that is?

hana-linode reviewed (commented):
Volume Create flow LGTM.

I'm not seeing the disabled regions for Linode Create though. I'm adding `currentCapability="Block Storage"` to the RegionSelect in `SelectRegionPanel`. Is there something I'm missing?

coliu-akamai commented:
hmm I'm actually seeing the same thing 😢 looking into it

abailly-akamai commented:
@hana-linode & @coliu-akamai 

Fixed a couple issues:
1. there was an issue with the memoization of regions which resulted in not seeing the proper disabled region when navigating the APP. 
2. i replaced miami with atlanta so our results are consistent
3. i fixed the css warning
Now, for 
- `SelectRegionPanel`: `currentCapability="Linodes"` => Singapore and Tokyo should be disabled
- `VolumeCreate`: `currentCapability="Block Storage"` => Atlanta and Tokyo should be disabled

hana-linode reviewed (approved):
- Confirmed keyboard accessibility
- RegionSelect disabled state for Volume Create and Linode Create
- No more console warning for styling
